packages/protocol/contracts/layer1/based/TaikoInbox.sol

```diff
@@ -90,94 +89,100 @@ abstract contract TaikoInbox is EssentialContract, ITaikoInbox, ITaiko {
-                    signalSlots: _paramsArray[i].signalSlots,
-                    anchorInput: _paramsArray[i].anchorInput,
-                    baseFeeConfig: config.baseFeeConfig
-                });
-            }
-
-            require(metas_[i].txListHash != 0, BlobNotFound());
-            bytes32 metaHash = keccak256(abi.encode(metas_[i]));
-
-            BlockV3 storage blk = state.blocks[stats2.numBlocks % config.blockRingBufferSize];
-            // SSTORE #1
-            blk.metaHash = metaHash;
-
-            // SSTORE #2 {{
-            blk.blockId = stats2.numBlocks;
-            blk.timestamp = updatedParams.timestamp;
-            blk.anchorBlockId = updatedParams.anchorBlockId;
-            blk.nextTransitionId = 1;
-            blk.verifiedTransitionId = 0;
-            // SSTORE }}
-
-            // Update lastBlock to reference the most recently proposed block.
-            lastBlock = BlockInfo(metaHash, updatedParams.timestamp, updatedParams.anchorBlockId);
+        require(calldataUsed || _batchParams.numBlobs != 0, BlobNotSpecified());
+
+        updatedParams = _validateBatchParams(
+            _batchParams,
+            config.maxAnchorHeightOffset,
+            config.maxSignalsToReceive,
+            config.maxBlocksPerBatch,
+            lastBatch
+        );
+
+        // This section constructs the metadata for the proposed batch, which is crucial for
+        // nodes/clients to process the batch. The metadata itself is not stored on-chain;
+        // instead, only its hash is kept.
+        // The metadata must be supplied as calldata prior to proving the batch, enabling the
+        // computation and verification of its integrity through the comparison of the metahash.
+        //
+        // Note that `difficulty` has been removed from the metadata. The client and prover must use
+        // the following approach to calculate a block's difficulty:
+        //  `keccak256(abi.encode("TAIKO_DIFFICULTY", block.number))`
+        meta_ = BatchMetadata({
+            txListHash: calldataUsed ? keccak256(_txList) : _calcTxListHash(_batchParams.numBlobs),
+            extraData: bytes32(uint256(config.baseFeeConfig.sharingPctg)),
+            coinbase: _coinbase,
+            batchId: stats2.numBatches,
+            gasLimit: config.blockMaxGasLimit,
+            timestamp: updatedParams.timestamp,
+            parentMetaHash: lastBatch.metaHash,
+            proposer: _proposer,
+            livenessBond: config.livenessBond,
+            proposedAt: uint64(block.timestamp),
+            proposedIn: uint64(block.number),
+            txListOffset: _batchParams.txListOffset,
+            txListSize: _batchParams.txListSize,
+            numBlobs: calldataUsed ? 0 : _batchParams.numBlobs,
+            anchorBlockId: updatedParams.anchorBlockId,
+            anchorBlockHash: blockhash(updatedParams.anchorBlockId),
+            signalSlots: _batchParams.signalSlots,
+            blocks: _batchParams.blocks,
```

Comment:
When calldata is used, shall we make sure that only one block will be proposed?

Reply:
No, using calldata or blobs, multiple blocks can be proposed per batch.